src/isc-dhcp/patch/0009-Support-for-dual-tor-scenario.patch

```diff
@@ -0,0 +1,437 @@
+-	   name matches the one specified in circuit_id. */
+-
+-	for (ip = interfaces; ip; ip = ip->next) {
+-		if (ip->circuit_id &&
+-		    ip->circuit_id_len == circuit_id_len &&
+-		    !memcmp(ip->circuit_id, circuit_id, circuit_id_len))
+-			break;
++	if (enable_support_for_dual_tor) {
++		/* If circuit _id does not match, leave ip as NULL. */
++		if ((*out)->circuit_id && (*out)->circuit_id_len == circuit_id_len &&
++			!memcmp((*out)->circuit_id, circuit_id, circuit_id_len))
++			ip = *out;
++	} else {
++		/* Scan the interface list looking for an interface whose
++		   name matches the one specified in circuit_id. */
++		for (ip = interfaces; ip; ip = ip->next) {
++			if (ip->circuit_id &&
++				ip->circuit_id_len == circuit_id_len &&
++				!memcmp(ip->circuit_id, circuit_id, circuit_id_len))
++				break;
++		}
+ 	}
+ 
+ 	/* If we got a match, use it. */
+@@ -1668,9 +1767,13 @@ add_relay_agent_options(struct interface_info *ip, struct dhcp_packet *packet,
+ 			*sp++ = 4u;
+ 			memcpy(sp, &giaddr.s_addr, 4);
+ 			sp += 4;
+-			packet->giaddr = uplink->addresses[0];
+-			log_debug ("Adding link selection suboption"
+-				   " with addr: %s", inet_ntoa(giaddr));
++			log_debug ("Adding link selection suboption with addr: %s", inet_ntoa(giaddr));
++
++			if (enable_support_for_dual_tor) {
++				packet->giaddr = loopback_addr;
++			} else {
++				packet->giaddr = uplink->addresses[0];
++			}
+ 		}
+ 
+ #ifdef RELAY_PORT
+@@ -2398,6 +2501,7 @@ void request_v4_interface(const char* name, int flags) {
+         struct interface_info *tmp = NULL;
+         int len = strlen(name);
+         isc_result_t status;
++        struct downstream_intf_list *ci = NULL;
+ 
+         if (len >= sizeof(tmp->name)) {
+                 log_fatal("%s: interface name too long (is %d)", name, len);
+@@ -2413,6 +2517,15 @@ void request_v4_interface(const char* name, int flags) {
+ 		  (flags & INTERFACE_UPSTREAM ? 'Y' : 'N'),
+ 		  (flags & INTERFACE_DOWNSTREAM ? 'Y' : 'N'));
+ 
++	if (flags & INTERFACE_DOWNSTREAM) {
++		ci = ((struct downstream_intf_list *)dmalloc(sizeof *ci, MDL));
++		if (!ci)
++			log_fatal("no memory for downstream interface pointer.\n");
++		ci->next = downstream_intfs;
++		downstream_intfs = ci;
++		ci->interface = tmp;
++	}
++
+         strncpy(tmp->name, name, len);
+         interface_snorf(tmp, (INTERFACE_REQUESTED | flags));
+         interface_dereference(&tmp, MDL);
+@@ -2487,3 +2600,74 @@ free_interface_alias_map(void) {
+ 	free(g_interface_name_alias_map);
+ 	g_interface_name_alias_map_size = 0;
+ }
++
++static void
++free_downstream_intfs(void) {
++	struct downstream_intf_list *cdi;
++	while (downstream_intfs) {
++		cdi = downstream_intfs;
++		downstream_intfs = downstream_intfs->next;
++		free(cdi);
++	}
++}
++
++static void *
++receive_pkt_from_loopback(void *arg) {
++	struct interface_info loopback_intf;
++	struct sockaddr_in address;
++	union {
++		unsigned char buffer [4095];
```

Comment:
Suggest changing name to `packbuf` and adding a comment for clarity to align with rest of codebase, e.g.,

```c
unsigned char packbuf [4095]; /* Packet input buffer.
                               * Must be as large as largest
                               * possible MTU. */
```

Reply:
the code is using the same underlying package's [nomenclature](https://salsa.debian.org/dhcp-team/isc-dhcp/-/blob/master/common/discover.c#L1094). When we use `-U` or similar approach, this will go away.